enhancements/microshift/generic-device-plugin-for-microshift.md

```diff
@@ -0,0 +1,461 @@
+---
+title: generic-device-plugin-for-microshift
+authors:
+  - pmtk
+reviewers:
+  - "@DanielFroehlich, MicroShift PM"
+  - "@ggiguash, MicroShift contributor"
+  - "@pacevedom, MicroShift contributor"
+  - "@eslutsky, MicroShift contributor"
+  - "@copejon, MicroShift contributor"
+approvers:
+  - "@jerpeter1, MicroShift Staff Engineer"
+api-approvers:
+  - none
+creation-date: 2025-05-13
+last-updated: 2025-05-13
+tracking-link:
+  - https://issues.redhat.com/browse/OCPSTRAT-2031
+---
+
+# Generic Device Plugin for MicroShift
+
+## Summary
+
+[Generic Device Plugin](https://github.com/squat/generic-device-plugin) is a
+solution for attaching devices not needing special initialization
+(e.g. serial, USB, certain video devices) to containers. The goal of this enhancement is to
+outline its adoption for MicroShift.
+
+## Motivation
+
+MicroShift is built for edge devices, which commonly use peripherals
+like sensors and adapters.
+Currently, MicroShift does not provide any solution to pass these peripherals
+through to Pods without requiring elevated privileges.
+We want to provide a solution for attaching simple devices (not requiring any special
+initialization) to Pods without potentially compromising on security.
+
+### User Stories
+
+* As a MicroShift device administrator, I want to attach simple devices to Pods
+  without elevating privileges (e.g. mounting `/dev`, adding extraneous capabilities).
+
+### Goals
+
+* Include functionality of
+  [Generic Device Plugin](https://github.com/squat/generic-device-plugin)
+  in MicroShift
+
+### Non-Goals
+
+* Adapt Generic Device Plugin (e.g. create Operator) for OpenShift or other Kubernetes distribution
+
+## Proposal
+
+* Import Generic Device Plugin code into MicroShift binary
+* Integrate the Generic Device Plugin's configuration into MicroShift config
+  and add a toggle option
+
+### Workflow Description
+
+**User** is a human user responsible for setting up and managing Edge Devices.
+
+1. The user enables the Generic Device Plugin using MicroShift configuration
+1. Optionally, the user modifies the configuration of Generic Device Plugin, 
+   for example changes device groups and paths
+1. The user (re)starts MicroShift
+1. MicroShift starts the Generic Device Plugin controller which creates plugin
+   goroutines for each device group.
+1. Plugin goroutines detect the devices and report them with the Kubelet
+1. Kubelet updates the Node's status with new devices for scheduler to be aware of Node's capabilities
+1. User deploys a workload utilizing some device
+   (could be at any point, it won't be scheduled/started until the device is available)
+
+In order for workload to get the device attached, the **container** specification
+must include the device in its limits. Below is an example of StatefulSet which
+references a device in a container's limits.
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
```

Comment:
Any specific reason for choosing statefulSet here ?

Reply:
Not really, it's what zigbee2mqtt uses and I was using it for testing PoC.

Reply:
@pmtk So when it comes to QE, does verifying it with statefulSet would be sufficient  ?

Reply:
Yes, it could be even done with just a `Pod`.
It could be done with deployment or daemonset, but it's also a bit about intent: deployments are rather ephemeral (like scaling up and down creates/deleted Pods randomly, but statefulset's pods keep sticky ID). DaemonSets are per node application, and while it gets job done on single node MicroShift, I think to use DaemonSet with a device, you'd have to have the same device on each of the nodes.